vault/cors.go

```diff
@@ -0,0 +1,109 @@
+package vault
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/logical"
+)
+
+var errCORSNotConfigured = errors.New("CORS is not configured")
+
+// CORSConfig stores the state of the CORS configuration.
+type CORSConfig struct {
+	sync.RWMutex
+	Enabled        bool     `json:"enabled"`
+	AllowedOrigins []string `json:"allowed_origins"`
+}
+
+func (c *Core) saveCORSConfig() error {
+	view := c.systemBarrierView.SubView("config/")
+
+	entry, err := logical.StorageEntryJSON("cors", c.corsConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create CORS config entry: %v", err)
+	}
+
+	if err := view.Put(entry); err != nil {
+		return fmt.Errorf("failed to save CORS config: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Core) loadCORSConfig() error {
+	view := c.systemBarrierView.SubView("config/")
+
+	// Load the config in
+	out, err := view.Get("cors")
+	if err != nil {
+		return fmt.Errorf("failed to read CORS config: %v", err)
+	}
+	if out == nil {
+		return nil
+	}
+
+	config := new(CORSConfig)
+	err = out.DecodeJSON(config)
+	if err != nil {
+		return err
+	}
+
+	c.corsConfig = config
+
+	return nil
+}
+
+// Enable takes either a '*' or a comma-seprated list of URLs that can make
+// cross-origin requests to Vault.
+func (c *CORSConfig) Enable(urls string) error {
```

Comment:
The input to this function will be a `[]string` based on the field data type change I've suggested in an another comment.